File: project-{topic}-enhanced/src/controllers/UserController.cpp
```cpp
#include "UserController.hpp"
#include "../logger/Logger.hpp"
#include "../utils/AppException.hpp"
#include "../middleware/AuthMiddleware.hpp" // To access user claims from request
#include "../utils/CryptoUtils.hpp" // For hashing new passwords

#include "crow.h"
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <vector>

UserController::UserController(UserService& userService)
    : userService(userService) {}

// Helper to get authenticated user ID and role
std::pair<int, std::string> getAuthenticatedUserClaims(const crow::request& req) {
    int userId = req.get_context<AuthMiddleware::Context>().user_id;
    std::string userRole = req.get_context<AuthMiddleware::Context>().user_role;
    return {userId, userRole};
}

crow::response UserController::getAuthenticatedUser(const crow::request& req) {
    try {
        auto [userId, userRole] = getAuthenticatedUserClaims(req);

        std::optional<User> userOpt = userService.findById(userId);
        if (!userOpt) {
            // This should ideally not happen if AuthMiddleware successfully verified a user ID
            Logger::error("UserController: Authenticated user {} not found in database.", userId);
            throw NotFoundException("Authenticated user not found.");
        }

        Logger::info("UserController: User {} retrieved their own profile.", userId);
        nlohmann::json resBody;
        resBody["status"] = "success";
        resBody["user"] = userOpt.value().toJson();
        // Remove sensitive fields like password_hash before sending
        resBody["user"].erase("password_hash");
        return crow::response(200, resBody.dump());

    } catch (const std::runtime_error& e) {
        // Catch exceptions from service layer and re-throw as AppExceptions
        throw InternalServerErrorException(e.what());
    }
}

crow::response UserController::updateAuthenticatedUser(const crow::request& req) {
    try {
        auto [userId, userRole] = getAuthenticatedUserClaims(req);
        nlohmann::json reqBody = nlohmann::json::parse(req.body);

        std::optional<User> existingUserOpt = userService.findById(userId);
        if (!existingUserOpt) {
            Logger::error("UserController: Authenticated user {} not found for update (DB inconsistency).", userId);
            throw NotFoundException("Authenticated user not found.");
        }
        User existingUser = existingUserOpt.value();

        bool updated = false;
        if (reqBody.contains("email")) {
            std::string newEmail = reqBody["email"].get<std::string>();
            if (!newEmail.empty() && newEmail != existingUser.getEmail()) {
                // Basic email format check (more robust needed for prod)
                if (newEmail.find('@') == std::string::npos) {
                    throw BadRequestException("Invalid email format.");
                }
                // Check if new email already exists for another user
                std::optional<User> userWithNewEmail = userService.findByEmail(newEmail);
                if (userWithNewEmail && userWithNewEmail->getId() != userId) {
                    throw ConflictException("Email already taken by another user.");
                }
                existingUser.setEmail(newEmail);
                updated = true;
            }
        }
        if (reqBody.contains("password")) {
            std::string newPassword = reqBody["password"].get<std::string>();
            if (newPassword.length() < 8) {
                throw BadRequestException("New password must be at least 8 characters long.");
            }
            std::string newHashedPassword = CryptoUtils::hashPassword(newPassword);
            existingUser.setPasswordHash(newHashedPassword);
            updated = true;
        }
        // Username and role cannot be changed by user themselves, only by admin for role.

        if (updated) {
            userService.updateUser(userId, existingUser);
            Logger::info("UserController: User {} updated their profile.", userId);
        } else {
            Logger::info("UserController: User {} made update request with no changes.", userId);
        }

        nlohmann::json resBody;
        resBody["status"] = "success";
        resBody["message"] = updated ? "User profile updated successfully" : "No changes applied.";
        return crow::response(200, resBody.dump());

    } catch (const nlohmann::json::parse_error& e) {
        throw BadRequestException("Invalid JSON in request body.");
    } catch (const std::runtime_error& e) {
        throw BadRequestException(e.what());
    }
}

crow::response UserController::deleteAuthenticatedUser(const crow::request& req) {
    try {
        auto [userId, userRole] = getAuthenticatedUserClaims(req);

        userService.deleteUser(userId);
        Logger::info("UserController: User {} deleted their account.", userId);

        nlohmann::json resBody;
        resBody["status"] = "success";
        resBody["message"] = "User account deleted successfully";
        return crow::response(200, resBody.dump());

    } catch (const std::runtime_error& e) {
        throw InternalServerErrorException(e.what());
    }
}

crow::response UserController::getAllUsers(const crow::request& req) {
    try {
        auto [userId, userRole] = getAuthenticatedUserClaims(req);

        // Authorization check: Only admin can view all users
        if (userRole != "admin") {
            Logger::warn("UserController: User {} (role: {}) attempted to access all users list.", userId, userRole);
            throw ForbiddenException("You are not authorized to view all users.");
        }

        std::vector<User> users = userService.getAllUsers();
        nlohmann::json userArray = nlohmann::json::array();
        for (const auto& user : users) {
            nlohmann::json userJson = user.toJson();
            userJson.erase("password_hash"); // Remove sensitive information
            userArray.push_back(userJson);
        }

        Logger::info("UserController: Admin user {} fetched all users.", userId);
        nlohmann::json resBody;
        resBody["status"] = "success";
        resBody["users"] = userArray;
        return crow::response(200, resBody.dump());

    } catch (const std::runtime_error& e) {
        throw InternalServerErrorException(e.what());
    }
}
```

File: project-{topic}-enhanced/controllers/userController.js
```javascript
const Joi = require('joi');
const userService = require('../services/userService');
const { AppError } = require('../utils/errorHandler');
const logger = require('../utils/logger');

const userIdSchema = Joi.object({
  id: Joi.string().uuid().required(),
});

const createUserSchema = Joi.object({
  username: Joi.string().alphanum().min(3).max(30).required(),
  email: Joi.string().email().required(),
  password: Joi.string().min(8).required(),
  role: Joi.string().valid('user', 'admin').default('user'),
});

const updateUserSchema = Joi.object({
  username: Joi.string().alphanum().min(3).max(30).optional(),
  email: Joi.string().email().optional(),
  role: Joi.string().valid('user', 'admin').optional(),
  password: Joi.string().min(8).optional(), // Allow password change, but handle hashing in service
}).min(1); // At least one field must be present for update

exports.getAllUsers = async (req, res, next) => {
  try {
    const users = await userService.findAllUsers();
    logger.info('Fetched all users.');
    res.status(200).json({
      status: 'success',
      results: users.length,
      data: {
        users: users.map(user => ({ id: user.id, username: user.username, email: user.email, role: user.role })),
      },
    });
  } catch (error) {
    next(error);
  }
};

exports.getUserById = async (req, res, next) => {
  try {
    const { error } = userIdSchema.validate(req.params);
    if (error) {
      throw new AppError(error.details[0].message, 400);
    }

    const user = await userService.findUserById(req.params.id);
    if (!user) {
      throw new AppError('User not found', 404);
    }
    logger.info(`Fetched user with ID: ${req.params.id}`);
    res.status(200).json({
      status: 'success',
      data: { user: { id: user.id, username: user.username, email: user.email, role: user.role } },
    });
  } catch (error) {
    next(error);
  }
};

exports.createUser = async (req, res, next) => {
  try {
    const { error, value } = createUserSchema.validate(req.body);
    if (error) {
      throw new AppError(error.details[0].message, 400);
    }

    const user = await userService.createUser(value);
    logger.info(`Created new user with ID: ${user.id}`);
    res.status(201).json({
      status: 'success',
      message: 'User created successfully',
      data: { user: { id: user.id, username: user.username, email: user.email, role: user.role } },
    });
  } catch (error) {
    next(error);
  }
};

exports.updateUser = async (req, res, next) => {
  try {
    const { error: idError } = userIdSchema.validate(req.params);
    if (idError) {
      throw new AppError(idError.details[0].message, 400);
    }

    const { error: bodyError, value } = updateUserSchema.validate(req.body);
    if (bodyError) {
      throw new AppError(bodyError.details[0].message, 400);
    }

    const updatedUser = await userService.updateUser(req.params.id, value);
    if (!updatedUser) {
      throw new AppError('User not found', 404);
    }
    logger.info(`Updated user with ID: ${req.params.id}`);
    res.status(200).json({
      status: 'success',
      message: 'User updated successfully',
      data: { user: { id: updatedUser.id, username: updatedUser.username, email: updatedUser.email, role: updatedUser.role } },
    });
  } catch (error) {
    next(error);
  }
};

exports.deleteUser = async (req, res, next) => {
  try {
    const { error } = userIdSchema.validate(req.params);
    if (error) {
      throw new AppError(error.details[0].message, 400);
    }

    const deleted = await userService.deleteUser(req.params.id);
    if (!deleted) {
      throw new AppError('User not found', 404);
    }
    logger.info(`Deleted user with ID: ${req.params.id}`);
    res.status(204).json({
      status: 'success',
      data: null,
    });
  } catch (error) {
    next(error);
  }
};
```